crates/block-producer/src/batch_builder/mod.rs

```diff
@@ -206,3 +213,84 @@ where
         Ok(())
     }
 }
+
+pub struct BatchProducer {
+    pub batch_interval: Duration,
+    pub workers: NonZeroUsize,
+    pub mempool: Arc<Mutex<Mempool>>,
+    pub tx_per_batch: usize,
+}
+
+type BatchResult = Result<BatchJobId, (BatchJobId, BuildBatchError)>;
+
+/// Wrapper around tokio's JoinSet that remains pending if the set is empty,
+/// instead of returning None.
+struct WorkerPool(tokio::task::JoinSet<BatchResult>);
+
+impl WorkerPool {
+    async fn join_next(&mut self) -> Result<BatchResult, tokio::task::JoinError> {
+        if self.0.is_empty() {
+            std::future::pending().await
+        } else {
+            // Cannot be None as its not empty.
+            self.0.join_next().await.unwrap()
+        }
+    }
+
+    fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    fn spawn(&mut self, id: BatchJobId, transactions: Vec<TransactionId>) {
+        self.0.spawn(async move {
+            todo!("Do actual work like aggregating transaction data");
```

Comment:
Similar to the above comment: batch building could also take 3 - 5 seconds. Let's artificially simulate this for now.

Reply:
I was wondering if we should also simulate failures. So something like sleep a random period, and then randomly fail a batch every now and then.

If we do want this, the failure rate should probably be a configurable parameter. Similar for the entire block.

Though this might annoy users if this causes a user tx to expire. And could look bad ito throughput; though I don't think that's an issue right now (?).

Reply:
If this is configurable and not too difficult to add, I don't mind doing it.